src/http-api.js

```diff
@@ -385,24 +385,89 @@ module.exports.MatrixHttpApi.prototype = {
         if (!queryParams) {
             queryParams = {};
         }
-        if (!queryParams.access_token) {
-            queryParams.access_token = this.opts.accessToken;
+        if (this.authorizationHeaderSupported === undefined ||
+                this.authorizationHeaderSupported) {
+            if (isFinite(opts)) {
+                // opts used to be localTimeoutMs
+                opts = {
+                    localTimeoutMs: opts,
+                };
+            }
+            if (!opts) {
+                opts = {};
+            }
+            if (!opts.headers) {
+                opts.headers = {};
+            }
+            if (!opts.headers.Authorization) {
+                if (queryParams.access_token) {
+                    opts.headers.Authorization = "Bearer " + queryParams.access_token;
+                    delete queryParams.access_token;
+                } else {
+                    opts.headers.Authorization = "Bearer " + this.opts.accessToken;
+                }
+            }
+            if (queryParams.access_token) {
+                delete queryParams.access_token;
+            }
+        } else {
+            if (!queryParams.access_token) {
+                queryParams.access_token = this.opts.accessToken;
+            }
         }
 
-        const request_promise = this.request(
+        const requestPromise = this.request(
             callback, method, path, queryParams, data, opts,
         );
 
         const self = this;
-        request_promise.catch(function(err) {
+
+        if (this.authorizationHeaderSupported === undefined) {
+            const defer = q.defer();
+            const returnPromise = defer.promise;
+            returnPromise.abort = requestPromise.abort;
+
+            requestPromise.then((resp) => {
+                self.authorizationHeaderSupported = true;
```

Comment:
First: Although Header-Support is implemented for a long time it was not allowed in terms of CORS. Is got implemented [recently](https://github.com/matrix-org/synapse/pull/2285) and there was no release since then.

Beside that: I understand the "false sense of security"-point. And I have no problem removing the promise-chain-stuff.
I can change it to be an constructor-set decision.

But especially in terms of CORS:    Where should such check be implemented?